### DISCUSSION TITLE
API permission name for self-hosted runners?

### DISCUSSION THREAD

**Discussion by skyzyx** [community/community#25611] [Enterprise]
<p>Apparently I have too many links to GitHub documentation. I’ve replaced <code>https</code> with <code>httpx</code> in the links below.</p>
<hr>
<ol>
<li>
<p>Creating a new “GitHub App” (not “OAuth App”) in GitHub Enterprise v3.0 (soon migrating to v3.1).</p>
</li>
<li>
<p>Trying to do it entirely over the API and explicitly NOT the UI, by creating an “app manifest” (httpx://docs.github.com/en/enterprise-server@3.0/developers/apps/building-github-apps/creating-a-github-app-from-a-manifest).</p>
</li>
<li>
<p>Everything I’ve read about permissions on <a href="http://docs.github.com" rel="noopener nofollow ugc">docs.github.com</a> ends up pointing over to httpx://docs.github.com/en/enterprise-server@3.0/rest/reference/permissions-required-for-github-apps, which <strong>does not</strong> include the specific values that can be used with the API.</p>
</li>
<li>
<p>On a GHE instance, there is a large list of permissions available at a URL with this pattern:</p>
<pre><code class="lang-nohighlight">https://{HOSTNAME}/organizations/{ORG}/settings/apps/{APP}/permissions
</code></pre>
</li>
<li>
<p>The specific permission I’m trying set says:</p>
<p><strong>Self-hosted runners</strong> <br>
View and manage Actions self-hosted runners available to an organization. <br>
Access: Read &amp; write</p>
</li>
<li>
<p>In the documentation (httpx://docs.github.com/en/enterprise-server@3.0/developers/apps/building-github-apps/creating-a-github-app-from-a-manifest#github-app-manifest-parameters) there is a parameter called <code>default_permissions</code>.</p>
<p>What is the identifier (key) to use for this permission, where the value is <code>write</code>?</p>
<p>I’ve tried:</p>
<ul>
<li>the documented <code>Self-hosted runners</code>
</li>
<li>the guess <code>self-hosted runners</code>
</li>
<li>the guess <code>self-hosted_runners</code>
</li>
<li>the guess <code>self_hosted_runners</code>
</li>
<li>the guess <code>selfhosted_runners</code>
</li>
<li>the guess <code>runners</code>
</li>
</ul>
<p>…but ultimately, the <em>actual values</em> which can be used here are (as far as I can tell after several hours of digging and guessing) <em>undocumented</em>.</p>
<p><code>actions:read</code> and <code>checks:read</code> appear to work. Those are <strong>also</strong> undocumented, but I was able to figure it out by looking at the URLs, making an educated guess, and testing.</p>
</li>
<li>
<p>All of the tutorials I can find on the internet, including those on <a href="http://docs.github.com" rel="noopener nofollow ugc">docs.github.com</a>, all walk you through creating a new GitHub app via the UI. I am very explicitly trying to do this over the API.</p>
</li>
</ol>
<p>Any tips? Have I missed something? Is this not available in GHE yet?</p>
<hr>
<p>Here is my app manifest, redacted.</p>
<pre><code class="lang-json">{
  "public": true,
  "name": "My app",
  "description": "My app's description.",
  "url": "https://github.example.com/my-org/my-repo",
  "redirect_url": "http://localhost:9876/register/redirect",
  "default_events": [],
  "default_permissions": {
    "actions": "read",
    "checks": "read",
    "runners": "write"
  },
  "hook_attributes": {
    "url": "",
    "active": false
  }
}
</code></pre>
<p>WITH the <code>"runners": "write"</code> line, the error message I receive says:</p>
<blockquote>
<p>Invalid GitHub App configuration <br>
The configuration does not appear to be a valid GitHub App manifest. <br>
× <strong>Error</strong> <strong>Default permission records resource is not included in the list</strong></p>
</blockquote>
<p>WITHOUT the <code>"runners": "write"</code> line, the submission is successful.</p>

**Comment by skyzyx** [comment, marked as answer]
<p>The GitHub team finally updated the documentation. The permission I was looking for was <code>organization_self_hosted_runners</code>.</p>

**Comment by icj217** [comment]
I too ran into this issue where it was unclear what the proper permissions were. I ended up creating the app by hand and adding the permissions I needed for it to function, then I used the GH API to see what permissions it had using this command:
```
gh api \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /apps/<slug>
```

I was specifically interested in the `custom_property_values` event, which ended up needing the `organization_custom_properties` and `repository_custom_properties` permissions. It looks like permission names can be plural and can also be prefixed with `organization_` or `repository_`. 

If only GitHub could provide a public list of these permissions.

**Reply by theELITEbes** [reply]
[](https://github.com/community/community/discussions/25611#discussioncomment-9833036)
